### PR TITLE
SDN-2599: [OVN-K]-4.9: Add sgw->lgw migration job at runtime

### PR DESCRIPTION
**Pull request by tssurya**
/cc trozet & @jluhrsen 

See https://docs.google.com/document/d/1SYaNQNW4sWVTE9XgHG6gKsS4ZActK8_Z7wILXU5YuoU/edit#heading=h.dqas96vg4g6m for details

### CODE REVIEW

**Comment by tssurya** [comment]
why are we timing out on the progressing command?

2022-01-06T14:51:51.921851436Z I0106 14:51:51.921777       1 log.go:184] Reconciling Network.operator.openshift.io cluster
2022-01-06T14:51:51.922774061Z I0106 14:51:51.922744       1 ovn_kubernetes.go:269] Overriding OVN configuration to &{GatewayMode:local EnableEgressIP:true DisableSNATMutlipleGWs:false}
2022-01-06T14:51:51.923050982Z I0106 14:51:51.923037       1 log.go:184] Starting render phase

```
2022-01-06T14:51:52.488770268Z I0106 14:51:52.488736       1 log.go:184] Set operator conditions:
2022-01-06T14:51:52.488770268Z - lastTransitionTime: "2022-01-06T14:27:41Z"
2022-01-06T14:51:52.488770268Z   status: "False"
2022-01-06T14:51:52.488770268Z   type: ManagementStateDegraded
2022-01-06T14:51:52.488770268Z - lastTransitionTime: "2022-01-06T14:27:41Z"
2022-01-06T14:51:52.488770268Z   status: "False"
2022-01-06T14:51:52.488770268Z   type: Degraded
2022-01-06T14:51:52.488770268Z - lastTransitionTime: "2022-01-06T14:27:41Z"
2022-01-06T14:51:52.488770268Z   status: "True"
2022-01-06T14:51:52.488770268Z   type: Upgradeable
2022-01-06T14:51:52.488770268Z - lastTransitionTime: "2022-01-06T14:51:52Z"
2022-01-06T14:51:52.488770268Z   message: DaemonSet "openshift-ovn-kubernetes/ovnkube-master" update is being processed
2022-01-06T14:51:52.488770268Z     (generation 2, observed generation 1)
2022-01-06T14:51:52.488770268Z   reason: Deploying
2022-01-06T14:51:52.488770268Z   status: "True"
2022-01-06T14:51:52.488770268Z   type: Progressing
2022-01-06T14:51:52.488770268Z - lastTransitionTime: "2022-01-06T14:30:27Z"
2022-01-06T14:51:52.488770268Z   status: "True"
2022-01-06T14:51:52.488770268Z   type: Available
```

**Comment by trozet** [comment]
Is this relevant:
```
Changing to local gateway mode
-------------------
gateway-mode.yaml
---------------------------------------------
apiVersion: v1
kind: ConfigMap
metadata:
    name: gateway-mode-config
    namespace: openshift-network-operator
data:
    mode: "local"
immutable: true
configmap/gateway-mode-config created
gateway-mode-config-map
---------------------------------------------
Name:         gateway-mode-config
Namespace:    openshift-network-operator
Labels:       <none>
Annotations:  <none>

Data
====
mode:
----
local
Events:  <none>
error: timed out waiting for the condition on clusteroperators/network
```
https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift_release/24985/rehearse-24985-pull-ci-openshift-ovn-kubernetes-release-4.9-e2e-aws-ovn-shared-to-local-gateway-mode-migration/1479090647435579392/artifacts/e2e-aws-ovn-shared-to-local-gateway-mode-migration/ovn-shared-to-local-gateway-mode-migration-4.9/build-log.txt

**Comment by trozet** [comment]
https://github.com/openshift/release/pull/24985/files#diff-c30fcd3e2e3528013ec5991e5b6c13eab2feca67036159a0db244e8470df1e69R30

300 seconds may not be sufficient to wait for rollout. When the configmap is created it can take a few minutes for CNO to realize the change and then another few minutes to update each pod. I think maybe you should bump this to 10 minutes.

**Comment by jluhrsen** [comment]
> https://github.com/openshift/release/pull/24985/files#diff-c30fcd3e2e3528013ec5991e5b6c13eab2feca67036159a0db244e8470df1e69R30
> 
> 300 seconds may not be sufficient to wait for rollout. When the configmap is created it can take a few minutes for CNO to realize the change and then another few minutes to update each pod. I think maybe you should bump this to 10 minutes.

Do we have to reboot nodes when rolling out this change?

I've spent a lot of time with an upgrade bug where I was rebooting nodes and I can tell you that it's no shorter than 45 seconds to reboot a node and that's after it's cordoned and drained. draining takes many seconds too. so for a 6 node cluster the bare min to just get all nodes rebooted will be close to 6m.

**Comment by tssurya** [comment]
> https://github.com/openshift/release/pull/24985/files#diff-c30fcd3e2e3528013ec5991e5b6c13eab2feca67036159a0db244e8470df1e69R30
> 
> 300 seconds may not be sufficient to wait for rollout. When the configmap is created it can take a few minutes for CNO to realize the change and then another few minutes to update each pod. I think maybe you should bump this to 10 minutes.

yes you got the error right: https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift_release/24985/rehearse-24985-pull-ci-openshift-ovn-kubernetes-release-4.9-e2e-aws-ovn-shared-to-local-gateway-mode-migration/1479090647435579392/artifacts/e2e-aws-ovn-shared-to-local-gateway-mode-migration/ovn-shared-to-local-gateway-mode-migration-4.9/build-log.txt

but the problem is not the daemonset roll-out, its the fact that its not detecting CNO moving to "Progressing=True" state so the command that is failing is https://github.com/openshift/release/pull/24985/files#diff-c30fcd3e2e3528013ec5991e5b6c13eab2feca67036159a0db244e8470df1e69R28
i.e:
`error: timed out waiting for the condition on clusteroperators/network`

The logic is, we create the config-map, then we wait for CNO to pick up this config-map (whether CNO's state moves to Progressing=True is what indicates if CNO has started rolling out the new comfig-map change and we are currently failing here), then we wait for ovnkube-node and ovnkube-master roll-outs. This worked fine on master actually. not sure why the 30sec window is not enough here since from the logs:

```
2022-01-06T14:51:51.921851436Z I0106 14:51:51.921777 1 log.go:184] Reconciling Network.operator.openshift.io cluster
2022-01-06T14:51:51.922774061Z I0106 14:51:51.922744 1 ovn_kubernetes.go:269] Overriding OVN configuration to &{GatewayMode:local EnableEgressIP:true DisableSNATMutlipleGWs:false}
2022-01-06T14:51:51.923050982Z I0106 14:51:51.923037 1 log.go:184] Starting render phase

2022-01-06T14:51:52.488770268Z I0106 14:51:52.488736       1 log.go:184] Set operator conditions:
2022-01-06T14:51:52.488770268Z - lastTransitionTime: "2022-01-06T14:27:41Z"
2022-01-06T14:51:52.488770268Z   status: "False"
2022-01-06T14:51:52.488770268Z   type: ManagementStateDegraded
2022-01-06T14:51:52.488770268Z - lastTransitionTime: "2022-01-06T14:27:41Z"
2022-01-06T14:51:52.488770268Z   status: "False"
2022-01-06T14:51:52.488770268Z   type: Degraded
2022-01-06T14:51:52.488770268Z - lastTransitionTime: "2022-01-06T14:27:41Z"
2022-01-06T14:51:52.488770268Z   status: "True"
2022-01-06T14:51:52.488770268Z   type: Upgradeable
2022-01-06T14:51:52.488770268Z - lastTransitionTime: "2022-01-06T14:51:52Z"
2022-01-06T14:51:52.488770268Z   message: DaemonSet "openshift-ovn-kubernetes/ovnkube-master" update is being processed
2022-01-06T14:51:52.488770268Z     (generation 2, observed generation 1)
2022-01-06T14:51:52.488770268Z   reason: Deploying
2022-01-06T14:51:52.488770268Z   status: "True"
2022-01-06T14:51:52.488770268Z   type: Progressing
2022-01-06T14:51:52.488770268Z - lastTransitionTime: "2022-01-06T14:30:27Z"
2022-01-06T14:51:52.488770268Z   status: "True"
2022-01-06T14:51:52.488770268Z   type: Available
```

It is clear CNO moves into PROGRESSING=True state within less than one second.

**Comment by tssurya** [comment]
> > https://github.com/openshift/release/pull/24985/files#diff-c30fcd3e2e3528013ec5991e5b6c13eab2feca67036159a0db244e8470df1e69R30
> > 300 seconds may not be sufficient to wait for rollout. When the configmap is created it can take a few minutes for CNO to realize the change and then another few minutes to update each pod. I think maybe you should bump this to 10 minutes.
> 
> Do we have to reboot nodes when rolling out this change?
> 
> I've spent a lot of time with an upgrade bug where I was rebooting nodes and I can tell you that it's no shorter than 45 seconds to reboot a node and that's after it's cordoned and drained. draining takes many seconds too. so for a 6 node cluster the bare min to just get all nodes rebooted will be close to 6m.

no this change doesn't need node reboots.

**Comment by tssurya** [comment]
ok I think we are good now: https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift_release/24985/rehearse-24985-pull-ci-openshift-ovn-kubernetes-release-4.9-e2e-aws-ovn-shared-to-local-gateway-mode-migration/1479385890928201728/artifacts/e2e-aws-ovn-shared-to-local-gateway-mode-migration/ovn-shared-to-local-gateway-mode-migration-4.9/build-log.txt

File: ci-operator/step-registry/ovn/shared-to-local-gateway-mode-migration-4.9/ovn-shared-to-local-gateway-mode-migration-4.9-commands.sh [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+
+# Giving upto 5mins for CNO to pick this up because each reconciliation loop is 5mins apart
+oc wait co network --for='condition=PROGRESSING=True' --timeout=300s
+# Wait until the ovn-kubernetes pods are restarted
+timeout 600s oc rollout status ds/ovnkube-node -n openshift-ovn-kubernetes
+timeout 600s oc rollout status ds/ovnkube-master -n openshift-ovn-kubernetes
+
+# ensure the gateway mode change was successful, if not no use proceeding with the test
+mode=$(oc get -n openshift-network-operator cm gateway-mode-config -o template --template '{{.data.mode}}')
+echo "Routing via host is set to ${mode}"
```
Comment (tssurya) [inline ci-operator/step-registry/ovn/shared-to-local-gateway-mode-migration-4.9/ovn-shared-to-local-gateway-mode-migration-4.9-commands.sh:36]: oops silly copy-paste mistake

**Comment by trozet** [comment]
/lgtm

thanks @tssurya !

**Pull request merged**